### PR TITLE
expression: clean code.

### PR DESCRIPTION
**Pull request by hanfei1991**
we have implemented all signatures of the function that can be pushed down

### CODE REVIEW

File: expression/distsql_builtin.go [hunk 1]
```diff
@@ -93,45 +92,7 @@ func newDistSQLFunction(sc *variable.StatementContext, exprType tipb.ExprType, a
 	// TODO: Too ugly...
```
Comment (shenli) [inline expression/distsql_builtin.go]: Can we remove this TODO?

Comment (zz-jason) [inline expression/distsql_builtin.go]: "TODO" can be removed ?

Comment (hanfei1991) [inline expression/distsql_builtin.go]: nonononono

Comment (hanfei1991) [inline expression/distsql_builtin.go]: no, this todo is for mock ctx.

Comment (zz-jason) [inline expression/distsql_builtin.go]: how about change the TODO message to a more specific one ?

File: expression/distsql_builtin.go [hunk 2]
```diff
@@ -93,45 +92,7 @@ func newDistSQLFunction(sc *variable.StatementContext, exprType tipb.ExprType, a
...
-			}
-		}
-	}
-	funcNode := &ast.FuncCallExpr{FnName: model.NewCIStr(funcName), Args: newArgs}
-	err := InferType(sc, funcNode)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	return funcNode.GetType(), nil
+	return NewFunction(ctx, name, types.NewFieldType(mysql.TypeUnspecified), args...)
```
Comment (shenli) [inline expression/distsql_builtin.go]: Why use TypeUnspecified?

File: expression/distsql_builtin.go (hunk 1 above)
Comment (hanfei1991) [inline expression/distsql_builtin.go]: ok

File: expression/distsql_builtin.go (hunk 2 above)
Comment (hanfei1991) [inline expression/distsql_builtin.go]: Let `NewFunction` to infer type

Comment (zz-jason) [inline expression/distsql_builtin.go]: @shenli it's a work around, after all expression refactored, this parameter should be removed from function `NewFunction`

**Comment by shenli** [comment]
LGTM

**Pull request merged**